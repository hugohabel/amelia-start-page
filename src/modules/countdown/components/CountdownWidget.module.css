.lightText {
  color: #FFFFFF;
}
.separatorWord {
  margin: 5px 10px;
}
.biggerText {
  font-size: 30px;
}