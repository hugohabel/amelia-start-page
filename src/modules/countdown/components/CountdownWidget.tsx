// Internal Dependencies
import { CountdownFormat } from '../../../common/types/amelia';
import {
  getDiffUntilDate,
  formatTimeDiff,
  isDateInThePast,
} from '../utils/timeCalculations/timeCalculations';
import { log } from '../../../common/utils/logs/logs';

// Styles
import styles from './CountdownWidget.module.css';
import sharedStyles from '../../../common/styles/Shared.module.css';

// Interfaces + Types
interface ICountdownWidgetProps {
  eventName: string;
  eventDate: string;
  eventFormat: CountdownFormat;
  eventEmoji: string;
}

/**
 * Countdown Widget
 *
 * @param params ICountdownWidgetProps
 * @returns JSX
 */
function CountdownWidget({ eventDate, eventName, eventFormat, eventEmoji }: ICountdownWidgetProps) {
  let isEventInThePast = false;
  let formattedOutput = '';

  try {
    const timeDiff = getDiffUntilDate(eventDate);
    isEventInThePast = isDateInThePast(timeDiff);
    formattedOutput = formatTimeDiff(timeDiff, eventFormat);
  } catch (err: any) {
    log({
      level: 'error',
      message: err.message,
      data: `eventDate: ${eventDate} - eventFormat: ${eventFormat}`,
      file: 'CountdownWidget.ts',
      location: 'CountdownWidget',
    });
  }

  return (
    <div className={styles.countdownWidget}>
      <h3>{`${eventEmoji} ${eventName}`}</h3>
      <p className={styles.separatorWord}>{!isEventInThePast && `in`}</p>
      <h2 className={sharedStyles.mainText}>{formattedOutput}</h2>
    </div>
  );
}

export { CountdownWidget };
