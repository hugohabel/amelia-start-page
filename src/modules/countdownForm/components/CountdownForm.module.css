.formError {
  font-size: 14px;
  color: #E13A38;
}