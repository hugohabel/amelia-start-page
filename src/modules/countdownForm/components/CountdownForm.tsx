// External Dependencies
import React from 'react';
import * as Yup from 'yup';
import { useFormik } from 'formik';
import { v4 as uuidv4 } from 'uuid';

// Internal Dependencies
import { useAppDispatch } from '../../../common/contexts/AppState';

// Styles
import styles from './CountdownForm.module.css';

function CountdownForm() {
  const dispatch = useAppDispatch();

  const formik = useFormik({
    initialValues: {
      eventName: '',
    },
    validationSchema: Yup.object({
      eventName: Yup.string()
        .min(2, 'Must be 2 characters or more')
        .required('Event name is required')
    }),
    onSubmit: values => {
      dispatch({
        type: 'addNewCountdownEvent',
        event: {
          id: uuidv4(),
          name: values.eventName,
          date: '2021-08-10T05:00:00.000Z',
          background: ''
        }
      });
      dispatch({
        type: 'setActiveView',
        activeView: null
      });
    },
  });

  return (
    <>
      <form onSubmit={formik.handleSubmit} autoComplete="off">
        <input 
          type="text" 
          id="eventName"
          name="eventName"
          placeholder="Event name"
          onChange={formik.handleChange}
          value={formik.values.eventName}
        />
        {formik.touched.eventName && formik.errors.eventName ? (
          <div className={styles.formError}>{formik.errors.eventName}</div>
        ) : null}
        
        <button type="submit">Submit</button>
      </form>
    </>
  );
}

export default CountdownForm;