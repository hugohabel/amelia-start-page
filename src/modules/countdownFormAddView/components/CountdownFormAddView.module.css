/* Go Back */

.backContainer {
  display: flex;
  margin-bottom: 16px;
  align-items: center;
  cursor: pointer;
}
.backContainer > svg {
  color: #2D3A99;
}
.backContainer > span {
  margin-left: 8px;
  font-size: 12px;
}
.backContainer:hover > svg {
  color: #FFA92B;
}

/* End Go Back */

/* Title + Subtitle */

.title {
  margin-bottom: 5px;
  font-weight: 600;
  color: #2D3A99;
  letter-spacing: -1;
}
.subtitle {
  margin-bottom: 20px;
  font-size: 12px;
}

/* Title + Subtitle */