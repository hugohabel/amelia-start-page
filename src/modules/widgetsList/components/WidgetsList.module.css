/* Widgets List */

.widgetsListItem {
  background: #FFFFFF;
  padding: 15px 25px;
  border-radius: .428rem;
  line-height: 20px;
  display: flex;
  cursor: pointer;
  margin-bottom: 20px;
  border: 1px solid transparent;
}
.widgetsListItem:hover {
  border: 1px solid #7367f0;
}
.widgetsListItem > svg {
  color: #F3AC4A;
  margin-right: 8px;
}
.widgetsListItemHidden {
  display: none;
}
.widgetsListItemUpcoming {
  color: #CCCCCC;
  cursor: auto;
}
.widgetsListItemUpcoming > svg {
  color: #FAD299;
}
.widgetsListItemUpcomingTitle {
  color: #7367f0;
  margin-left: 10px;
  background: #FFEACA;
  padding: 0px 8px;
  border-radius: 5px;
  font-size: 12px;
  font-weight: bold;
}
.widgetsListTitle {
  margin-bottom: 5px;
  font-weight: 600;
  color: #7367f0;
  letter-spacing: -1;
}
.widgetsListSubtitle {
  margin-bottom: 20px;
  font-size: 12px;
}

/* End Widgets List */