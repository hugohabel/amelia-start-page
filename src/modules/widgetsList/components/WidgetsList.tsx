// External Dependencies
import HourglassEmptyOutlinedIcon from '@material-ui/icons/HourglassEmptyOutlined';
import WbSunnyOutlinedIcon from '@material-ui/icons/WbSunnyOutlined';
import WatchLaterOutlinedIcon from '@material-ui/icons/WatchLaterOutlined';

// Internal Dependencies
import { WIDGETS_LIST } from '../../../common/constants';
import { useAppDispatch } from '../../../common/contexts/AppState';
import { TActiveView } from '../../../common/types/AppState';

// Styles
import styles from './WidgetsList.module.css';

// Interfaces + Types
type TWidgetsIcons = {
  // eslint-disable-next-line no-undef
  [key: string]: JSX.Element;
};

const widgetsIcons: TWidgetsIcons = {
  countdown: <HourglassEmptyOutlinedIcon fontSize="small" />,
  weather: <WbSunnyOutlinedIcon fontSize="small" />,
  timezone: <WatchLaterOutlinedIcon fontSize="small" />,
};

function WidgetsList() {
  const dispatch = useAppDispatch();

  // Handler Functions
  function handleListItemClick(widgetName: string) {
    dispatch({
      type: 'setActiveView',
      activeView: `add_${widgetName}` as TActiveView,
    });
  }

  return (
    <div className={styles.widgetsList}>
      <h2 className={styles.widgetsListTitle}>New widget</h2>
      <p className={styles.widgetsListSubtitle}>Select a widget from the list below</p>
      {WIDGETS_LIST.map((widget) => {
        const widgetsListItemClasses =
          // eslint-disable-next-line no-nested-ternary
          widget.status === 1
            ? `${styles.widgetsListItem}`
            : widget.status === -1
            ? `${styles.widgetsListItemHidden}`
            : `${styles.widgetsListItem} ${styles.widgetsListItemUpcoming}`;

        return (
          <div
            data-testid={widget.name}
            className={widgetsListItemClasses}
            key={widget.name}
            onClick={() => {
              if (widget.status === 1) {
                handleListItemClick(widget.name);
              }
            }}>
            {widgetsIcons[widget.name]}
            <span>{widget.title}</span>
            {widget.status === 0 && (
              <span className={styles.widgetsListItemUpcomingTitle}>Upcoming!</span>
            )}
          </div>
        );
      })}
    </div>
  );
}

export { WidgetsList };
