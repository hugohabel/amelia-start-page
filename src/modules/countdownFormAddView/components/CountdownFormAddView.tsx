// External Dependencies
import ArrowBackOutlinedIcon from '@material-ui/icons/ArrowBackOutlined';

// Internal Dependencies
import { useAppDispatch } from '../../../common/contexts/AppState';
import CountdownForm from '../../countdownForm/components/CountdownForm';

// Styles
import styles from './CountdownFormAddView.module.css';

function CountdownFormAddView() {
  const dispatch = useAppDispatch();

  // Handler Functions
  function handleGoBackClick() {
    dispatch({
      type: 'setActiveView',
      activeView: null,
    });
  }

  return (
    <>
      {/* Go Back */}
      <div data-testid="go-back" className={styles.backContainer} onClick={handleGoBackClick}>
        <ArrowBackOutlinedIcon fontSize="small" />
        <span>Go Back</span>
      </div>
      {/* End Go Back */}

      {/* Title */}
      <h2 className={styles.title}>New countdown</h2>
      <p className={styles.subtitle}>Add the details for your countdown widget</p>
      {/* End Title */}

      {/* Countdown Form */}
      <CountdownForm />
      {/* End Countdown Form */}
    </>
  );
}

export default CountdownFormAddView;
