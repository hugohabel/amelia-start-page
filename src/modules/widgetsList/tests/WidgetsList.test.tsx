// External Dependencies
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

// Internal Dependencies
import { AppStateProvider, useAppDispatch } from '../../../common/contexts/AppState';
import { WidgetsList } from '../components/WidgetsList';

// Mocked Functions
jest.mock('../../../common/contexts/AppState', () => ({
  ...(jest.requireActual('../../../common/contexts/AppState') as object),
  useAppDispatch: jest.fn(),
}));
let mockAppDispatch: ReturnType<typeof useAppDispatch>;

describe('Widgets List', () => {
  it('matches snapshot', async () => {
    const component = render(<WidgetsList />);

    expect(component.asFragment()).toMatchSnapshot();
  });

  describe('Widgets List - Handlers', () => {
    beforeEach(() => {
      mockAppDispatch = jest.fn();
      (useAppDispatch as jest.Mock).mockReturnValue(mockAppDispatch);
    });

    it('triggers state updates when clicking on the widgets list item', async () => {
      const component = render(
        <AppStateProvider>
          <WidgetsList />
        </AppStateProvider>
      );

      // Get Go Back Element
      const countdownItemElement = await component.findByTestId('countdown');
      expect(countdownItemElement).toBeInTheDocument();

      // Trigger Click
      fireEvent.click(countdownItemElement);

      await waitFor(() => {
        expect(mockAppDispatch).toHaveBeenCalledWith({
          type: 'setActiveView',
          activeView: 'add_countdown',
        });
      });
    });
  });
});
