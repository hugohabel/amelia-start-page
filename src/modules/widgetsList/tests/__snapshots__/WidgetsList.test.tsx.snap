// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widgets List matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="widgetsList"
  >
    <h2
      class="widgetsListTitle"
    >
      New widget
    </h2>
    <p
      class="widgetsListSubtitle"
    >
      Select a widget from the list below
    </p>
    <div
      class="widgetsListItem"
      data-testid="countdown"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 2v6h.01L6 8.01 10 12l-4 4 .01.01H6V22h12v-5.99h-.01L18 16l-4-4 4-3.99-.01-.01H18V2H6zm10 14.5V20H8v-3.5l4-4 4 4zm-4-5l-4-4V4h8v3.5l-4 4z"
        />
      </svg>
      <span>
        Countdown
      </span>
    </div>
    <div
      class="widgetsListItem widgetsListItemUpcoming"
      data-testid="weather"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6.76 4.84l-1.8-1.79-1.41 1.41 1.79 1.79zM1 10.5h3v2H1zM11 .55h2V3.5h-2zm8.04 2.495l1.408 1.407-1.79 1.79-1.407-1.408zm-1.8 15.115l1.79 1.8 1.41-1.41-1.8-1.79zM20 10.5h3v2h-3zm-8-5c-3.31 0-6 2.69-6 6s2.69 6 6 6 6-2.69 6-6-2.69-6-6-6zm0 10c-2.21 0-4-1.79-4-4s1.79-4 4-4 4 1.79 4 4-1.79 4-4 4zm-1 4h2v2.95h-2zm-7.45-.96l1.41 1.41 1.79-1.8-1.41-1.41z"
        />
      </svg>
      <span>
        Weather
      </span>
      <span
        class="widgetsListItemUpcomingTitle"
      >
        Upcoming!
      </span>
    </div>
    <div
      class="widgetsListItemHidden"
      data-testid="timezone"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm.5-13H11v6l5.2 3.2.8-1.3-4.5-2.7V7z"
        />
      </svg>
      <span>
        Timezone Converter
      </span>
    </div>
  </div>
</DocumentFragment>
`;
